src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -116,13 +124,22 @@ protected override ContextPropagationToken CreatePropagationTokenCore(ContextPro
 
         protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
         {
+            // Headers can only be written once. Throw on subsequent call to write response header instead of silent no-op.
+            if (HttpContext.Response.HasStarted)
+            {
+                throw new InvalidOperationException("Response headers can only be sent once per call.");
```

Comment:
```suggestion
                throw new InvalidOperationException("Response headers have already been sent. Response headers can only be sent once per call.");
```

Reply:
I'm using the same error message as the Grpc.Core implementation.

Reply:
Hmm ok. I think this is more clear but we could explain more in logging later.